rooms API 예외 처리, Entity 검증, Pagination 모듈 도입, DB 관계 개선

yxxjxxeee opened the pull request:
## #️⃣ 연관된 이슈
- #8 

## 📝 작업 내용
> **예외 처리 통일**: ValidationPipe의 exceptionFactory 적용 및 ExceptionFilter로 예외 응답 통일
**Room Entity 생성 및 검증 로직 응집**: 비밀번호, 인원수 검증 로직 Entity 내부로 이동, 기존 Service 레이어 검증 제거
**Pagination 모듈 도입**: PaginationService 생성, Room 모듈 findAll에 적용
**DB 관계 및 onDelete 옵션 개선**: 테이블 관계 재정의 및 onDelete 명확화, UserStats 테이블 새로 추가, Room 물리 삭제 전환

yxxjxxeee committed:
refactor: (#8) exceptionFactory 적용 및 ExceptionFilter로 예외 응답 통일

yxxjxxeee committed:
Merge remote-tracking branch 'origin' into refactor/#8/rooms-api

yxxjxxeee committed:
refactor: (#8) Entity 생성 및 검증 로직 추가, 예외 필터 타입 개선

- Room Entity에 규칙 검증 로직 추가
- AllExceptionsFilter에서 any 타입 제거, 타입 안전한 처리 적용

yxxjxxeee committed:
refactor: (#8) Pagination 모듈 도입 및 Rooms 모듈 구조 개선

- getRooms에서 PaginationService를 활용해 커서 기반 페이지네이션 적용
- RoomsService에서 ResRoomDto 반환
- Room 엔티티의 DTO 의존성을 제거하고 개별 값 기반으로 동작 가능하도록 개선
- RoomsRepository는 Prisma 처리만 담당하도록 책임 분리

yxxjxxeee committed:
refactor: (#8) onDelete 옵션 개선, UserStats 추가 및 물리 삭제 적용

hobiJeong commented on src/common/pagination/pagination.service.ts:
```diff
@@ -0,0 +1,23 @@
+    if (cursor) {
+      const index = items.findIndex((item) => item.id === cursor);
+      startIndex = index >= 0 ? index + 1 : 0;
+    }
```
```suggestion
    const startIndex = (() => {
      if (cursor) {
        const index = items.findIndex((item) => item.id === cursor)
        return index >= 0 ? index + 1 : 0
      } else {
        return 0
      }
    })()
```
혹시 let이 싫다면 이렇게 IIFE를 이용하는 방법도 있습니다

rrgks6221 commented on prisma/migrations/20250923144128_add_room_fields/migration.sql:
프로젝트 초반이고 아직 개발자 로컬환경밖에 갖춰지지 않아서 큰 문제는 없을거로 보이는데요
마이그레이션은 기본적으로 실행된후에는 수정 삭제되면 안됩니다
변경이 필요하다면 새로운 마이그레이션 파일을 생성하도록 하는게 바람직해보여요
mvp_init도 삭제 -> schama.prisma 파일을 변경 -> 마이그레이션 파일을 새로 생성 하신거같은데 이름이 같더라도 새로 생성하는 습관이 생겼으면 좋겠습니다.

1. 이미 마이그레이션 된 항목이 유실됨
2. 언제 어떤 스크립트에 의해서 변경됐는지 확인이 불가능
3. 백업 및 롤백도 불가능해짐

이미 작업된 부분에서 다른 팀원분 로컬 환경에서도 문제가 없다면 당장에는 인지만 하시고 꼭 고치지 않으셔도 됩니다

yxxjxxeee replied:
이해했습니다! 프로젝트 초반이라 전체 구조를 크게 수정한 김에 기존 마이그레이션 파일을 초기화하고 새로 mvp_init을 만드는 게 더 깔끔하다고 생각해서 그렇게 진행했습니다. 
다만 말씀처럼 추후에는 기록이 유실되면 백업에 문제가 생기니까 앞으로는 기존 마이그레이션은 그대로 두고 새로운 파일을 추가하는 방식으로 진행하겠습니다.

rrgks6221 commented on prisma/schema.prisma:
```diff
@@ -55,8 +55,11 @@ model User {
   providerId     String?   @map("provider_id") @db.VarChar(50)
   createdAt      DateTime  @default(now()) @map("created_at")
   updatedAt      DateTime  @updatedAt @map("updated_at")
+  isDeleted      Boolean   @default(false) @map("is_deleted")
```
이 필드는 아마 삭제여부에 대한 필터링할 때 사용될거같은데 맞을까요??
맞다면 deletedAt(DATE TYPE)에 대한 is null 검사보다 boolean 검사가 필터링에 대해 더 빨라서 추가로 저장하는 것일까요??(성능 차이는 저도 몰라서 여쭤봐요)

yxxjxxeee replied:
맞습니다! 원래 삭제 여부를 필터링하기 위해 isDeleted와 deletedAt을 사용하려고 했습니다.
그런데 다시 보니 현재 구조에서는 물리 삭제를 사용해도 될 것 같아 이 필드는 삭제할 계획입니다.
그리고 일반적으로 boolean 필드(isDeleted)로 필터링하는 것이 deletedAt IS NULL 검사보다 조금 더 빠르고 코드도 더 직관적인 것 같아서 그렇게 설계했습니다.

rrgks6221 commented on src/rooms/dtos/requests/query-room.dto.ts:
```diff
@@ -17,6 +12,16 @@ export class QueryRoomDto {
 
   @IsOptional()
-  @IsEnum(RoomStatus, { message: '상태는 WAITING 또는 IN_PROGRESS 중 하나여야 합니다.' })
+  @IsBoolean({ message: '비공개 여부는 true 또는 false이어야 합니다.' })
+  isPrivate?: boolean;
```
쿼리스트링에 사용될 dto로 보이는데 'true' | 'false' 같은 값을 boolean으로 변환하는 과정이 추가되면 좋을거같아요

yxxjxxeee replied:
현재 main.ts에서 ValidationPipe를 transform: true로 글로벌 설정해 두어서
쿼리스트링으로 들어오는  'true' | 'false'  문자열이 자동으로 boolean으로 변환됩니다!

rrgks6221 replied:
혹시 이거 테스트해보셨을까요??
라이브러리 버전에 따라서 달라질 여지는 있겠지만 아마도 지원하지 않을 기능일거같아서요

`transform: true` 설정은 request로 들어오는 plain object를 class Object로 변환해주는 기능이예요. class-transfomer는 class object여야만 작동하는 데코레이터라 `transform: true` 옵션을 줘야지만 변환이 적용되기 시작하구요

class-validator의 IsBoolean에서 암시적으로 변환해주는것을 기대할수도있겠는데 대부분의 class-validator의 기능에서는 라이브러리 이름에 맞게 검증만 할 뿐 타입변환은 해주지 않습니다

```ts
// plain object
const plain = {};
console.log(plain); // {}

// class object
class CObject {
}
const cObject = new CObject();
console.log(cObject); // CObject {}
```

yxxjxxeee replied:
제가 잘못된 정보를 알고 있었네요!
테스트해보니까 말씀해주신 대로 `transform: true`는 plain object를 DTO 인스턴스로 변환해주는 역할만 하고 실제로 `'true'` | `'false'` 문자열을 boolean으로 바꿔주는 건 따로 `@Transform()` 같은 처리가 필요하겠네요.
알려주셔서 감사합니다 앞으로는 테스트를 좀 더 꼼꼼하게 해봐야겠어요 ㅠㅠ

hobiJeong replied:
transformOptions: {
        enableImplicitConversion: true,
      },
ValidationPipe의 해당 옵션이 암시적 변환을 해주는 것으로 알긴 하는데 권장하진 않는 걸로 압니다.
사용하더라도 boolean으로의 변환이 예상한대로 동작하는지 체크가 필요할 것 같습니다

yxxjxxeee replied:
실제로 enableImplicitConversion 옵션이 있군요! 테스트해보니 동작은 확인됐습니다.
다만 boolean 변환 이슈가 있을 수 있다고 하셔서 사용은 안 하고 명시적으로 처리하는 방향으로 갈 것 같습니다.
알려주셔서 감사합니다!

rrgks6221 commented on src/rooms/entities/rooms.entity.ts:
```diff
@@ -0,0 +1,80 @@
+import { RoomMode } from '@prisma/client';
+import { BadRequestException } from '@nestjs/common';
+
+export class Room {
```
👍️

rrgks6221 commented on src/rooms/entities/rooms.entity.ts:
```diff
@@ -0,0 +1,80 @@
+      isPrivate: this.isPrivate,
+      hashedPassword,
+    };
+  }
```
바깥(데이터베이스)에 대한 객체 생성 로직을 응집한 부분 너무 좋습니다.
하나 사소한 부분에서 fromPrisma같은 네이밍은 엔티티(가장 안쪽에 있는 영역으로 외부에 대한 존재를 알 필요가 없음)가 prisma의 존재를 알고 의존하는거처럼 보일 수 있는 부분이 있어요 (이 부분은 변수명에 관련돼있고 이해하기 어려운 부분이라 두셔도 됩니다)

rrgks6221 commented on src/rooms/rooms.service.ts:
```diff
@@ -1,101 +1,95 @@
 import { PasswordEncoderService } from 'src/common/hashing/password-encoder.service';
-import { RoomInterface, RoomQueryInterface } from './rooms.interface';
-import { RoomMode } from '@prisma/client';
+import { PaginationService } from 'src/common/pagination/pagination.service';
+import { PaginatedResult } from 'src/common/pagination/pagination.interface';
+import { FindRoomsQuery } from './rooms.interface';
+import { Room } from './entities/rooms.entity';
```
roomService의 리턴 객체가 prismaRoom에서 resRoom으로 변경됐는데 res에 관련된 처리는 controller에서 하는게 어떤가 싶어요
prismaRoom이 controller에 넘어가는게 걸리신다면 entity를 리턴해줘도 됩니다.

service 레이어에서 DTO를 리턴해야하냐 entity를 리턴해야하냐 말이 많은데 개인적으로는 entity를 리턴하는게 재사용측면에서 더 좋더라고요

kimjiwonpg98 replied:
저도 동일한 생각입니다
entity를  순수 db 엔티티로 볼꺼냐
도메인 객체로 볼꺼냐 로 갈리는거같은데

개인적으로 재사용측면도 그렇고 더 편합니다

yxxjxxeee replied:
저도 그 부분에서 계속 고민했었는데 말씀하신 것처럼 컨트롤러에서 res 관련 처리를 담당하도록 리팩토링해보겠습니다!

rrgks6221 commented on src/common/pagination/pagination.service.ts:
```diff
@@ -0,0 +1,23 @@
+import { PaginatedResult } from './pagination.interface';
+
+@Injectable()
+export class PaginationService {
```
Service로 하지 않고 PaginationDto같은 클래스로 만들어서 생성자만으로 사용하는건 어떻게 생각하시나요??

yxxjxxeee replied:
좋은 제안인 것 같습니다. 현재 PaginationService가 순수 함수만 제공하고 있어서 DI까지 사용하는 것이 조금 과한 것 같아요.
클래스의 static 메서드로 만들거나 별도의 유틸리티 함수로 분리하면 더 간단하고 명확할 것 같습니다.
그 방향으로 수정해보겠습니다!

Kimsoo0119 commented on src/rooms/rooms.service.ts:
```diff
@@ -1,101 +1,95 @@
     };
-
-    return this.roomsRepository.create(roomData);
   }
```
현재 페이지네이션이 전체 데이터를 조회한 뒤 slice로 필요한 범위를 잘라내는 방식으로 이해했습니다.
다만, 데이터가 많아질수록 성능 저하가 발생할 수 있을 것 같습니다.

Prisma에서 제공하는 count로 전체 개수를 계산하고
cursor와 take를 활용해 필요한 범위만 조회하는 방식도 고려해볼 수 있을 것 같습니다.

yxxjxxeee replied:
말씀하신 대로 현재는 전체 데이터를 조회 후 slice로 자르는 방식을 사용하고 있습니다.
다만 데이터 규모가 아직 크지 않아 이 방식으로도 성능상 큰 문제가 없다고 판단했습니다.
추후 데이터가 많아질 경우를 대비해 제안 주신 방향 참고해서 cursor와 take 기반 페이지네이션으로 리팩토링을 고려해 보겠습니다.

yxxjxxeee committed:
refactor: (#8) DTO 변환, 응답 처리, 페이지네이션, 엔티티 메서드명 개선

- user 테이블에서 delete 관련 필드 제거
- DTO 변환(transform) 로직 추가
- 컨트롤러에서 응답(resDto) 변환 처리 개선
- 페이지네이션 기능 클래스로 변경
- 엔티티 메서드명 변경

yxxjxxeee committed:
Merge branch 'develop' into refactor/#8/rooms-api
